server/channels/app/post.go

```diff
@@ -797,27 +792,36 @@ func (a *App) publishWebsocketEventForPermalinkPost(c request.CTX, post *model.P
 	}
 
 	permalinkPreviewedPost := post.GetPreviewPost()
-	for _, cm := range channelMembers {
-		if permalinkPreviewedPost != nil {
-			post.Metadata.Embeds[0].Data = permalinkPreviewedPost
-		}
-
-		postForUser := a.sanitizePostMetadataForUserAndChannel(c, post, permalinkPreviewedPost, permalinkPreviewedChannel, cm.UserId)
+	if permalinkPreviewedPost != nil {
+		post.Metadata.Embeds[0].Data = permalinkPreviewedPost
+	} else {
+		return false, nil
+	}
 
-		// Using DeepCopy here to avoid a race condition
-		// between publishing the event and setting the "post" data value below.
-		messageCopy := message.DeepCopy()
-		broadcastCopy := messageCopy.GetBroadcast()
-		broadcastCopy.UserId = cm.UserId
-		messageCopy.SetBroadcast(broadcastCopy)
+	postCopy := post.Clone()
 
-		postJSON, jsonErr := postForUser.ToJSON()
-		if jsonErr != nil {
-			mlog.Warn("Failed to encode post to JSON", mlog.Err(jsonErr))
+	if permalinkPreviewedChannel != nil && !a.HasPermissionToReadChannel(c, post.UserId, permalinkPreviewedChannel) {
+		post.Metadata.Embeds[0].Data = nil
+	}
+	// Using DeepCopy here to avoid a race condition
+	// between publishing the event and setting the "post" data value below.
+	messageCopy := message.DeepCopy()
+	broadcastCopy := messageCopy.GetBroadcast()
+	broadcastCopy.ChannelHook = func(userID string, ev *model.WebSocketEvent) bool {
+		if permalinkPreviewedChannel != nil && !a.HasPermissionToReadChannel(c, userID, permalinkPreviewedChannel) {
+			postCopy.Metadata.Embeds[0].Data = nil
+		} else {
+			return false
 		}
-		messageCopy.Add("post", postJSON)
-		a.Publish(messageCopy)
+		postJSON, err := postCopy.ToJSON()
+		if err != nil {
+			mlog.Warn("Failed to encode post to JSON", mlog.Err(err))
+		}
+		ev.AddWithCopy("post", postJSON)
```

Comment:
> and we never set post on messageCopy to begin with.

We do. In the higher layers.

Reply:
> We do. In the higher layers.

Eek. So previously we unambiguously set the `post` property from within this method before publishing, but now it's required that the caller do this otherwise this method won't work correctly for some users.

Reply:
I think the expectation for this method is that - if the post contains a permalink, then it will do whatever is necessary to make changes to the websocket event and publish it. But if it does not, then publishing falls on the higher layer itself. With that logic, I think it makes sense that the caller already set the `post` property before calling. Because if there's no permalink, the caller has to publish the post.

Reply:
> I think the expectation for this method is that - if the post contains a permalink, then it will do whatever is necessary to make changes to the websocket event and publish it. But if it does not, then publishing falls on the higher layer itself. With that logic, I think it makes sense that the caller already set the post property before calling. Because if there's no permalink, the caller has to publish the post.

I think the coupling is still wrong -- here's a (modified) implementation that matches your expectations of the caller but would be broken in the case I described:
```go
	published, err := a.publishWebsocketEventForPermalinkPost(c, rpost, message)
	if err != nil {
		return nil, err
	}
	if !published {
		message.Add("post", postJSON)
		a.Publish(message)
	}